private/pkg/command/process.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package command
+
+import (
+	"context"
+	"errors"
+	"os/exec"
+
+	"go.uber.org/multierr"
+)
+
+type process struct {
+	cmd        *exec.Cmd
+	runner     *runner
+	terminated bool
+}
+
+func newProcess(cmd *exec.Cmd, runner *runner) *process {
+	return &process{
+		cmd:    cmd,
+		runner: runner,
+	}
+}
+
+func (p *process) Run(ctx context.Context) error {
+	if err := p.Start(); err != nil {
+		return err
+	}
+	return p.Wait(ctx)
+}
+
+func (p *process) Start() error {
+	err := p.cmd.Start()
+	if err == nil {
+		p.runner.incement()
+	}
+	return err
+}
+
+func (p *process) Wait(ctx context.Context) error {
+	// There will not be a second call to wait. Always decrement process
+	// counters when we return.
+	if p.terminated {
+		return errors.New("process already terminated")
+	}
+	p.terminated = true
+	defer p.runner.decrement()
```

Comment:
If a caller doesn't call Wait, Runner will be in deadlock. The semaphore needs to be decremented regardless of the call to Wait

Reply:
Calling Run calls Wait. It's still the same foolproof execution and collection of a ephemeral process.

The caller manages to the call to Wait if they use Start. There's no way around that because the use case is to execute a process that lives beyond the calling function's return.

So you're right, if the caller doesn't collect the process with Wait, then the process remains accounted for. That's expected behavior.

Reply:
In the current implementation, a user is required to call `Process.Wait`. A user could call `Process.Start`, throw away the `Process` object, and never call `Process.Wait`, which will cause a deadlock as more and more `Process.Start` calls are made. This isn't acceptable - a user should not have to call `Process.Wait` for a deadlock not to occur.

You can deal with this by starting a goroutine in `Process.Start` that watches on `*exec.Cmd.Wait`, and decrements when it returns.

The goal of this package is to make it so that users are unable to mess up - this package, as currently implemented, lets users cause a major issue if they do not correctly use `Process`.

Reply:
Additionally, a user could call `Process.Run` and `Process.Start` on the same `Process` object, whereas since there was no stateful object returned in the previous API, it was not possible to misuse any returned stateful object. The API as it stands leaves more room for error for a critical function

Reply:
I can setup a watcher routine as it will help in some cases of a forgotten close. It may produce some other odd behaviors (e.g. getting io.ErrClosedPipe) but the caller was broken anyway.

Managing a process is fundamentally stateful (hello halting problem, how I missed you). It's a problem even with the current command package: all you have to do is pass a background context (which are the majority being passed around) and leave open a stdin pipe if the program takes one. In fact, using a background context and calling Run in a goroutine was one of the ways I was going to create a long-lived process, but that felt a little hacky. I can still pursue that route if having command return a process handle of any kind feels worse.

For reference on why I'm looking at this kind of functionality, check out the latest draft PR for building modules directly from bare git storage: #2018

Reply:
We must limit the number of concurrent background processes, and any solution where the user has to call `Process.Wait` to prevent a potentially-system-side deadlock isn't acceptable